src/egraph.rs

```diff
@@ -202,6 +216,51 @@ impl<L: Language> EGraph<L> {
         pruned
     }
 
+    pub fn fold_constants(&mut self) -> usize {
+        let mut folded = 0;
+        let mut class_new_node = HashMap::default();
+        // finds a foldable expression for each class.
+        // if any, evalute it and add to map
+        for (id, class) in &self.classes {
+            // scan the nodes for foldable expr.
+            for node in &class.nodes {
+                match node {
+                    Expr::Operator(op, cids) => {
+                        let mut consts = Vec::new();
+                        let mut all_consts = true;
+                        // check if each child is a constant
+                        for cid in cids {
+                            let c = &self.get_eclass(*cid);
+                            if let Some(const_e) = c.get_constant() {
+                                consts.push(const_e);
+                            } else {
+                                all_consts = false;
+                                break;
+                            }
+                        }
+                        // evaluate if all children are constant
+                        if all_consts {
+                            let ne = Expr::Operator(op.clone(), consts);
+                            let v = L::eval(&RecExpr::from(ne));
+                            let const_e = Expr::Constant(v);
+                            class_new_node.insert(*id, const_e);
+                            break;
+                        }
+                    }
+                    _ => {}
+                }
+            }
+        }
+        // add and merge the new folded constants
+        for (cid, new_node) in class_new_node {
+            let aid = self.add(new_node).id;
+            self.union(cid, aid);
+            folded += 1;
```

Comment:
declare folded down here, closer to use.

Reply:
Also, only add to folded if the result of the add wasn't there. Otherwise you can't tell if `fold_constants` hits fixpoint without pruning.